File: src/file-history/restore.ts
import { readdir, readFile, stat, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { getFileHistoryDir } from "../config/paths.ts";

export interface BackupEntry {
	timestamp: number;
	path: string;
	size: number;
}

/**
 * List backups for a file, sorted newest-first.
 */
export async function listBackups(filePath: string, projectPath?: string): Promise<BackupEntry[]> {
	const histDir = getFileHistoryDir(projectPath, filePath);
	let files: string[];
	try {
		files = await readdir(histDir);
	} catch {
		return [];
	}

	const entries: BackupEntry[] = [];
	for (const f of files) {
		if (!f.endsWith(".bak")) continue;
		const ts = Number.parseInt(f.replace(".bak", ""), 10);
		if (Number.isNaN(ts)) continue;
		const fullPath = join(histDir, f);
		const info = await stat(fullPath);
		entries.push({ timestamp: ts, path: fullPath, size: info.size });
	}

	return entries.sort((a, b) => b.timestamp - a.timestamp);
}

/**
 * Restore a backup to the original file path.
 */
export async function restoreBackup(filePath: string, timestamp: number, projectPath?: string): Promise<string> {
	const histDir = getFileHistoryDir(projectPath, filePath);
	const backupPath = join(histDir, `${timestamp}.bak`);

	let content: string;
	try {
		content = await readFile(backupPath, "utf-8");
	} catch {
		return `Error: Backup not found for timestamp ${timestamp}`;
	}

	await writeFile(filePath, content, "utf-8");
	return `Restored ${filePath} from backup ${timestamp}`;
}

File: src/file_history/restore.rs
//! List backups and restore a specific version.

use std::path::{Path, PathBuf};

use regex::Regex;

use super::meta::FileHistoryMeta;
use crate::config::paths::get_file_history_dir;

#[derive(Debug, Clone)]
pub struct BackupEntry {
    pub version: u32,
    pub path: PathBuf,
    pub size: u64,
}

pub fn list_backups(file_path: &Path, project_path: Option<&str>) -> Vec<BackupEntry> {
    let base_dir = get_file_history_dir(project_path);
    let mut meta = FileHistoryMeta::new(&base_dir);
    let entry = match meta.find_by_path(file_path) {
        Some(e) => e,
        None => return Vec::new(),
    };
    let uuid_dir = base_dir.join(&entry.uuid);
    let read = match std::fs::read_dir(&uuid_dir) {
        Ok(r) => r,
        Err(_) => return Vec::new(),
    };
    let re = Regex::new(r"^v(\d+)\.bak$").unwrap();
    let mut entries: Vec<BackupEntry> = read
        .flatten()
        .filter_map(|e| {
            let name = e.file_name().to_string_lossy().into_owned();
            let caps = re.captures(&name)?;
            let version: u32 = caps.get(1)?.as_str().parse().ok()?;
            let path = e.path();
            let size = e.metadata().ok()?.len();
            Some(BackupEntry {
                version,
                path,
                size,
            })
        })
        .collect();
    entries.sort_by(|a, b| b.version.cmp(&a.version));
    entries
}

pub fn restore_backup(file_path: &Path, version: u32, project_path: Option<&str>) -> String {
    let base_dir = get_file_history_dir(project_path);
    let mut meta = FileHistoryMeta::new(&base_dir);
    let entry = match meta.find_by_path(file_path) {
        Some(e) => e,
        None => return format!("Error: No backup history found for {}", file_path.display()),
    };
    let backup_path = base_dir.join(&entry.uuid).join(format!("v{version}.bak"));
    let content = match std::fs::read(&backup_path) {
        Ok(c) => c,
        Err(_) => {
            return format!(
                "Error: Backup version {version} not found for {}",
                file_path.display()
            )
        }
    };
    if let Err(e) = std::fs::write(file_path, content) {
        return format!("Error: {e}");
    }
    format!("Restored {} from backup v{version}", file_path.display())
}
